hamilton/src/main.cpp

```diff
@@ -0,0 +1,303 @@
+                     const std::vector<std::string>& joint_names, double time_delay);
+
+/**
+ * Sends a ROS trajectory to the robot controller
+ */
+bool executeTrajectory(const trajectory_msgs::JointTrajectory& trajectory);
+
+/**
+ * Plans for them, appends in overall_robot_traj
+ */
+void addFreeSpaceSegment(std::vector<geometry_msgs::Pose> waypoints, moveit::planning_interface::MoveGroup& group, robot_trajectory::RobotTrajectory& overall_robot_traj);
+//second argument to avoid global variables
+
+/**
+ * Plans for process path points, and appends them to overall_robot_traj
+ * TODO make argument for planner generic => Use pathPlannerBase
+ */
+void addProcessSegment(TrajectoryVec points, descartes_planner::DensePlanner& planner, moveit::planning_interface::MoveGroup& group, descartes_core::RobotModelPtr model, robot_trajectory::RobotTrajectory& overall_robot_traj, ros::NodeHandle& nh);
+
+int main(int argc, char** argv)
+{
+  // Initialize ROS
+  ros::init(argc, argv, "descartes_tutorial");
+  ros::NodeHandle nh;
+
+  // Required for communication with moveit components
+  ros::AsyncSpinner spinner (1);
+  spinner.start();
+
+  // Initialize a MoveIt! group
+  moveit::planning_interface::MoveGroup group("manipulator");
+
+  // Move to home
+  group.setNamedTarget("home");
+  group.move();
+    
+  // Move to target pose
+  geometry_msgs::Pose target_pose;
+  target_pose.position.x = 0.25;
+  target_pose.position.y = 0.25;
+  target_pose.position.z = 1.0;
+  group.setPoseTarget(target_pose);
+  group.move();
+  
+
+  // Create a Descartes robot model and initialize it
+  descartes_core::RobotModelPtr model (new descartes_moveit::MoveitStateAdapter);
+
+  // Name of description on parameter server. Typically just "robot_description".
+  const std::string robot_description = "robot_description";
+
+  // name of the kinematic group you defined when running MoveitSetupAssistant
+  // TODO Fermi has an issue of hardcoded manipulator. Avoid in Hamilton
+  const std::string group_name = "manipulator";
+
+  // Name of frame in which you are expressing poses. Typically "world_frame" or "base_link".
+  const std::string world_frame = "base_link";
+
+  // tool center point frame (name of link associated with tool)
+  const std::string tcp_frame = "tool0";
+
+  if (!model->initialize(robot_description, group_name, world_frame, tcp_frame))
+  {
+    ROS_INFO("Could not initialize robot model");
+    return -1;
+  }
+
+  descartes_planner::DensePlanner planner;
+  planner.initialize(model);
+  robot_trajectory::RobotTrajectory overall_robot_traj(group.getCurrentState()->getRobotModel(), "manipulator"); //TODO group.getRobotModel
+
+
+
+  // Generate some MoveIt | free space waypoints
+  std::vector<geometry_msgs::Pose> waypoints;
+  for(int i = 0; i < 8; ++i)
+  {
+    target_pose.position.z += 0.05;
+    waypoints.push_back(target_pose); 
+  }
+  addFreeSpaceSegment(waypoints, group, overall_robot_traj);
```

Comment:
The approach you are taking here for defining free motion paths assumes that more that two points (start and end) will be provided.  In most cases the user or calling code will provide the end points only and leave it to the planner to figure out how to move in between.  Thus it be more appropriate to send a [MotionPlanningRequest](http://docs.ros.org/api/moveit_msgs/html/msg/MotionPlanRequest.html) to moveit so that we get to leverage the motion planning capabilities of the active planner (usually OMPL).  Furthermore, using a moveit planner will allow defining either cartesian or joint targets.